### ISSUE TITLE
ifsck false positives

### ISSUE THREAD

**Issue by kript** [irods/irods#4217]
### iRODS Version, OS and Version

4.1.12
Ubuntu 12.04

### What did you try to do?

run `ifsck` on a resource in a composite tree to check for missing/incorrect files. 

### Expected behavior

report any files not consistent in size on disk with the catalog.

### Observed behavior (including steps to reproduce, if applicable)

`ifsck -R "root;replicate;green;green1;irods-seq-sr01-ddn-ra08-15-16-17" -r /irods-seq-sr01-ddn-ra08-15-16-17`
...
`ERROR chkObjConsistency: rcGenQuery failed: status [-808000] genQueryOut [0x2006d50]`
for every file (at least, I suspect, I stopped it after a few hundred)

This appears to be a catalog lookup issue as investigating one of the files showed it to be identical (unless I've missed something) on disk as the info in the catalog;
```
ERROR chkObjConsistency: rcGenQuery failed: status [-808000] genQueryOut [0x2006d50] file [/irods-seq-sr01-ddn-ra08-15-16-17/21833/21833_1#229_quality_error.txt]
...
$ ls -lah /irods-seq-sr01-ddn-ra08-15-16-17/21833/21833_1#229_quality_error.txt
-rw------- 1 irods irods 46K Feb  9  2017 /irods-seq-sr01-ddn-ra08-15-16-17/21833/21833_1#229_quality_error.txt
irods@irods-seq-sr01:~$ ls -l /irods-seq-sr01-ddn-ra08-15-16-17/21833/21833_1#229_quality_error.txt
-rw------- 1 irods irods 46174 Feb  9  2017 /irods-seq-sr01-ddn-ra08-15-16-17/21833/21833_1#229_quality_error.txt

$ md5sum /irods-seq-sr01-ddn-ra08-15-16-17/21833/21833_1#229_quality_error.txt
83da3c134ac5851236f7abfca910ba7d  /irods-seq-sr01-ddn-ra08-15-16-17/21833/21833_1#229_quality_error.txt

$ ils -L /seq/21833/21833_1#229_quality_error.txt
  srpipe            0 root;replicate;seq-red;red4;irods-seq-i21-fg        46174 2017-02-09.16:28 & 21833_1#229_quality_error.txt
    83da3c134ac5851236f7abfca910ba7d    generic    /irods-seq-i21-fg/21833/21833_1#229_quality_error.txt
  srpipe            1 root;replicate;seq-green;green1;irods-seq-sr01-ddn-ra08-15-16-17        46174 2017-02-09.16:28 & 21833_1#229_quality_error.txt
    83da3c134ac5851236f7abfca910ba7d    generic    /irods-seq-sr01-ddn-ra08-15-16-17/21833/21833_1#229_quality_error.txt
```

There were no errors on the iRES, or on its Catalog server at the time the command was run.

What did I do wrong?

**Comment by alanking** [comment]
Is the hierarchy provided to `-R` correct? The hierarchy provided reads `root;replicate;green;...`, but the replicas of the example data object reside on `root;replicate;seq-green;...` or `root;replicate;seq-red;...`. Should `-R` be given `root;replicate;seq-green;...`?

I get the output you're seeing when I run `ifsck` against a physical directory with the wrong resource:
```
$ ils -Lr
/tempZone/home/rods:
  C- /tempZone/home/rods/foodir
/tempZone/home/rods/foodir:
  rods              0 pt_demo;demoResc        21769 2018-12-11.11:14 & foo
        generic    /var/lib/irods/Vault/home/rods/foodir/foo
  rods              1 pt_ufs;ufs0        21769 2018-12-11.11:14 & foo
        generic    /tmp/irods/ufs0vault/home/rods/foodir/foo
$ ifsck -R 'pt_demo;demoResc' -r /var/lib/irods/Vault/home/rods/foodir
$ ifsck -R 'pt_demo;demoResc' -r /tmp/irods/ufs0vault/home/rods/foodir
ERROR chkObjConsistency: rcGenQuery failed: status [-808000] genQueryOut [0x9b81e0] file [/tmp/irods/ufs0vault/home/rods/foodir/foo]
$ ifsck -R 'pt_ufs;ufs0' -r /tmp/irods/ufs0vault/home/rods/foodir
```

**Comment by kript** [comment]
🤦‍♂️ I think you might be onto something there.. I'll check tomorrow and let you know...

**Comment by kript** [comment]
Progress! Of a sort... I now get far fewer genQueryOut lines, which look like they're reporting something to check, but its not clear what from the output;

```
...
ERROR chkObjConsistency: rcGenQuery failed: status [-808000] genQueryOut [0x19e59b0] file [/irods-seq-sr01-ddn-ra08-15-16-17/20933/20933_1#270.cram]
# count how many such results we got from running the ifsck across the resource
$ wc -l ifsck_irods-seq-sr01-ddn-ra08-15-16-17_20181212
274 ifsck_irods-seq-sr01-ddn-ra08-15-16-17_20181212
#We have more file than that in the resource
$ iadmin lr irods-seq-sr01-ddn-ra08-15-16-17 | grep count
resc_objcount: 825561
# what does iRODS think the file should be
$ ils -L /seq/20933/20933_1#270.cram
  srpipe            0 root;replicate;seq-red;red3;irods-seq-i18-de     39371205 2017-11-06.16:57 & 20933_1#270.cram
    d18fefe23f978050e1447069e33c28f3    generic    /irods-seq-i18-de/20933/20933_1#270.cram
  srpipe            1 root;replicate;seq-green;green2;irods-seq-sr02-ddn-rd10-36-37-38     39371205 2017-11-06.16:58 & 20933_1#270.cram
    d18fefe23f978050e1447069e33c28f3    generic    /irods-seq-sr02-ddn-rd10-36-37-38/20933/20933_1#270.cram
#what is the local checksum on disk
$ md5sum /irods-seq-sr01-ddn-ra08-15-16-17/20933/20933_1#270.cram
d41d8cd98f00b204e9800998ecf8427e  /irods-seq-sr01-ddn-ra08-15-16-17/20933/20933_1#270.cram
# what is the local size on disk
$ ls -lah !$
ls -lah /irods-seq-sr01-ddn-ra08-15-16-17/20933/20933_1#270.cram
-rw------- 1 irods irods 0 Jun 29  2017 /irods-seq-sr01-ddn-ra08-15-16-17/20933/20933_1#270.cram
#checking a different file gave a different (also corrupt) response;
...
ERROR chkObjConsistency: rcGenQuery failed: status [-808000] genQueryOut [0x19e9af0] file [/irods-seq-sr01-ddn-ra08-15-16-17/17935/RunInfo.xml]
...
$ ils -L /seq/17935/RunInfo.xml
  srpipe            0 root;replicate;seq-green;green2;irods-seq-sr02-ddn-rd10-12-13-14          826 2015-10-27.15:41 & RunInfo.xml
    82f2b8cb380d5938737d778bd18186f4    generic    /irods-seq-sr02-ddn-rd10-12-13-14/17935/RunInfo.xml
  srpipe            1 root;replicate;seq-red;red2;irods-seq-i08-de          826 2015-10-27.15:41 & RunInfo.xml
    82f2b8cb380d5938737d778bd18186f4    generic    /irods-seq-i08-de/17935/RunInfo.xml
$ md5sum /irods-seq-sr01-ddn-ra08-15-16-17/17935/RunInfo.xml
82f2b8cb380d5938737d778bd18186f4  /irods-seq-sr01-ddn-ra08-15-16-17/17935/RunInfo.xml
$ ls -l !$
ls -l /irods-seq-sr01-ddn-ra08-15-16-17/17935/RunInfo.xml
-rw------- 1 irods irods 826 Oct 23  2015 /irods-seq-sr01-ddn-ra08-15-16-17/17935/RunInfo.xml
```
It would be helpful if the tool stated what it thought the discrepancy was.

**Comment by alanking** [comment]
The error code being returned (-808000) means `CAT_NO_ROWS_FOUND`. Since it is the same error code as before and that indicated a mismatch of resource hierarchy to physical path, I'm wondering if the issue is that some files in the physical path(s) provided to `ifsck` are not registered in iRODS. Could you try running an `iscan` on these directories?

Are you seeing other output which suggests legitimate issues in the integrity of the data, or has it mainly been genquery errors?

**Comment by kript** [comment]
I think you’re right about this; the files on disk didn’t correspond to their iRODS locations - I’ll try an iscan tomorrow.

It was all genquery errors. This filesystem was last mostly filled five years ago in the 3.3.1 days, and we find the odd disparity here and there (zero length replicas and the like).



On 12 December 2018 at 13:34:16, Alan King (notifications@github.com) wrote:

The error code being returned (-808000) means CAT_NO_ROWS_FOUND. Since it is the same error code as before and that indicated a mismatch of resource hierarchy to physical path, I'm wondering if the issue is that some files in the physical path(s) provided to ifsck are not registered in iRODS. Could you try running an iscan on these directories?

Are you seeing other output which suggests legitimate issues in the integrity of the data, or has it mainly been genquery errors?

**Comment by kript** [comment]
Running iscan seems to report the same files as not being registered in iRODS, so yay!

**Comment by trel** [comment]
As per the 'yay' above, the title of this issue appears to be... not true?

What, if any, action items should remain for this issue?   Slightly more information in the original output about what didn't match?  Is that even possible?

If there is nothing to do, we may be able to close this one.

**Comment by kript** [comment]
So... `ifsck` complains about files it finds on disk that are not present in iRODS _on that resource_ (but often, for us, are present elsewhere, as complete uploads).

The "yay" refers to the fact that to make the `ifsck` useful, we therefore need to run `iscan` first, fix any files mentioned as un-registered there (which turn out to be in the tens of thousands, across all the systems), before running `ifsck` so that it no longer tries to check files its shouldn't know about (from a catalog perspective), as I understand it ;

iscan - verify filesystem contents against catalog
ifsck - verify catalog contents against filesystem

In other words, `ifsck` is reporting files that I would expect `iscan` to complain against (or more accurately returning `CAT_NO_ROWS_FOUND`, which only become clear when `iscan` is then run), rather than taking the list of files that the _catalog_ believes are in that resources and checking for presence, size, etc etc...

So I think the action would be to pass a list of files the Catalog believes are on that resource, and then check them rather than its current behaviour , where it highlights files that are not registered in the catalog, but not in a helpful way. Perhaps the two tools could/should be combined?

**Comment by trel** [comment]
You can target `ifsck` to a particular resource hierarchy with `-R`.

```
To check files registered to the leaf of a resource hierarchy,
use the -R option with a full resource hierarchy (semicolon delimited):
  ifsck -R 'pt1;r1;s1' /localdir/full/path/to/file.txt
```
```
 -R  rescHier - if provided, will match replicas on this leaf resource,
     rather than using the hostname of the root of each file's resolved hierarchy.
```

`ifsck` - checks physical files for presence and size/checksum consistency in the iRODS catalog

`iscan` - checks either
 - logical paths for corresponding presence on disk(s) or
 - physical paths for presence in catalog

**Comment by kript** [comment]
right, so both tools check for presence of physical files - in my reading of the usage, only `iscan` should do this, `ifsck` should only verify files the catalog already knows about?

**Comment by trel** [comment]
Oh, I see...

You're saying running `ifsck` on a physical path that contains a file that is not registered in the catalog should **skip/not report** the 'extra' file on disk.  As a consistency check, it shouldn't care about files that are not already in the catalog.  That's `iscan`'s job.

aka...

the first error below should be reported, and the second one should not.

```
$ ifsck -r /var/lib/irods/Vault/home/rods/
CORRUPTION: local file [/var/lib/irods/Vault/home/rods/foo] size [84] not consistent with iRODS object [/tempZone/home/rods/foo] size [82].
ERROR chkObjConsistency: rcGenQuery failed: status [-808000] genQueryOut [0x2696d00] file [/var/lib/irods/Vault/home/rods/newstuff.txt]
```

You would like to see, instead...
```
$ ifsck -r /var/lib/irods/Vault/home/rods/
CORRUPTION: local file [/var/lib/irods/Vault/home/rods/foo] size [84] not consistent with iRODS object [/tempZone/home/rods/foo] size [82].
```

If this is correct, that seems... reasonable.   This is a change in behavior and would definitely need to be documented well.  But I do think it makes sense.

Please confirm my understanding - and we'll change the title of this issue.

**Comment by kript** [comment]
Yes, that's it exactly, thank you. I guess we could retain the existing functionality with a verbose flag, as having more than one way to do things is never a bad thing...
